app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -1,13 +1,17 @@
 class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
-  # You should configure your model like this:
-  # devise :omniauthable, omniauth_providers: [:twitter]
+  after_action :verify_authorized, except: [:new, :create, :destroy, :google_oauth2]
 
-  # You should also create an action method in this controller like this:
-  # def twitter
-  # end
-
-  # More info at:
-  # https://github.com/plataformatec/devise#omniauth
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+    
+    if @user.persisted?
```

Comment:
I see what you are doing here and I like the idea of returning a user object from the User class under any flow control situation, the strong typing ideals in my head like that, but it seems awkward to check for persistence. I think it might be more ruby-ish to just return nil from the user class, then you can just raise the Pundit::NotAuthorized error in the case they are not authorized to be logging in. `raise Pundit::NotAuthorizedError unless @user`. I'll leave a note in the user method too.